src/setup_payload/python/SetupPayload.py

```diff
@@ -0,0 +1,220 @@
+
+import Base38
+import click
+from bitarray import bitarray
+from bitarray.util import int2ba, zeros
+from construct import BitsInteger, BitStruct, Enum
+from stdnum.verhoeff import calc_check_digit
+
+# Format for constructing manualcode
+manualcode_format = BitStruct(
+    'version' / BitsInteger(1),
+    'vid_pid_present' / BitsInteger(1),
+    'discriminator' / BitsInteger(4),
+    'pincode_lsb' / BitsInteger(14),
+    'pincode_msb' / BitsInteger(13),
+    'vid' / BitsInteger(16),
+    'pid' / BitsInteger(16),
+    'padding' / BitsInteger(7),  # this is intentional as BitStruct only takes 8-bit aligned data
+)
+
+# Format for constructing qrcode
+# qrcode bytes are packed as lsb....msb, hence the order is reversed
+qrcode_format = BitStruct(
+    'padding' / BitsInteger(4),
+    'pincode' / BitsInteger(27),
+    'discriminator' / BitsInteger(12),
+    'discovery' / BitsInteger(8),
+    'flow' / Enum(BitsInteger(2),
+                  Standard=0, UserIntent=1, Custom=2),
+    'pid' / BitsInteger(16),
+    'vid' / BitsInteger(16),
+    'version' / BitsInteger(3),
+)
+
+
+class CommissioningFlow(enum.IntEnum):
+    Standard = 0,
+    UserIntent = 1,
+    Custom = 2
+
+
+class SetupPayload:
+    def __init__(self, discriminator, pincode, rendezvous=4, flow=CommissioningFlow.Standard, vid=0, pid=0):
+        self.long_discriminator = discriminator
+        self.short_discriminator = discriminator >> 8
+        self.pincode = pincode
+        self.discovery = rendezvous
+        self.flow = flow
+        self.vid = vid
+        self.pid = pid
+
+    def p_print(self):
+        print('{:<{}} :{}'.format('Flow', 24, self.flow))
+        print('{:<{}} :{}'.format('Pincode', 24, self.pincode))
+        print('{:<{}} :{}'.format('Short Discriminator', 24, self.short_discriminator))
+        if self.long_discriminator:
+            print('{:<{}} :{}'.format('Long Discriminator', 24, self.long_discriminator))
+        if self.discovery:
+            print('{:<{}} :{}'.format('Discovery Capabilities', 24, self.discovery))
+        if self.vid is not None and self.pid is not None:
+            print('{:<{}} :{:<{}} (0x{:04x})'.format('Vendor Id', 24, self.vid, 6, self.vid))
+            print('{:<{}} :{:<{}} (0x{:04x})'.format('Product Id', 24, self.pid, 6, self.pid))
+
+    def qrcode_dict(self):
+        return {
+            'version': 0,
+            'vid': self.vid,
+            'pid': self.pid,
+            'flow': int(self.flow),
+            'discovery': self.discovery,
+            'discriminator': self.long_discriminator,
+            'pincode': self.pincode,
+            'padding': 0,
+        }
+
+    def manualcode_dict(self):
+        return {
+            'version': 0,
+            'vid_pid_present': 0 if self.flow == CommissioningFlow.Standard else 1,
+            'discriminator': self.short_discriminator,
+            'pincode_lsb': self.pincode & 0x3FFF,    # 14 ls-bits
+            'pincode_msb': self.pincode >> 14,       # 13 ms-bits
```

Comment:
Why as lsb/msb? Suggest having a member that is the "passcode"

Also suggest using @dataclass (Data classes) for these (as a follow-up) and use dataclasses.asdict(x) to dict-ify them for dicts versions.

Reply:
Raised https://github.com/project-chip/connectedhomeip/issues/33549 for follow-up